cognite_toolkit/_cdf_tk/resource_classes/container_field_definitions.py

```diff
@@ -0,0 +1,361 @@
+        if index_type not in _INDEX_DEFINITION_CLASS_BY_TYPE:
+            raise ValueError(
+                f"invalid index type '{index_type}'. Expected one of {humanize_collection(_INDEX_DEFINITION_CLASS_BY_TYPE.keys(), bind_word='or')}"
+            )
+        cls_ = _INDEX_DEFINITION_CLASS_BY_TYPE[index_type]
+        return cast(Self, cls_.model_validate(data))
+
+    @model_serializer(mode="wrap", when_used="always", return_type=dict)
+    def serialize_index(self, handler: SerializerFunctionWrapHandler) -> dict:
+        serialized_data = handler(self)
+        return serialized_data
+
+
+class BtreeIndex(IndexDefinition):
+    _index_type = "btree"
+    index_type: Literal["btree"] = "btree"
+    by_space: bool | None = Field(default=None, description="Whether to make the index space-specific.")
+    cursorable: bool | None = Field(
+        default=None, description="Whether the index can be used for cursor-based pagination."
+    )
+
+
+class InvertedIndex(IndexDefinition):
+    _index_type = "inverted"
+    index_type: Literal["inverted"] = "inverted"
+
+
+_INDEX_DEFINITION_CLASS_BY_TYPE: MappingProxyType[str, type[IndexDefinition]] = MappingProxyType(
+    {c._index_type: c for c in IndexDefinition.__subclasses__()}
+)
+
+
+class PropertyTypeDefinition(BaseModelResource):
+    _property_type: ClassVar[str]
+    type: str
+
+    @model_validator(mode="wrap")
+    @classmethod
+    def find_property_type_cls(cls, data: Any, handler: ModelWrapValidatorHandler[Self]) -> Self:
+        if isinstance(data, PropertyTypeDefinition):
+            return cast(Self, data)
+        if not isinstance(data, dict):
+            raise ValueError(f"Invalid property type data '{type(data)}' expected dict")
+
+        if cls is not PropertyTypeDefinition:
+            return handler(data)
+
+        property_type = data.get("type")
+        if property_type is None:
+            raise ValueError("Missing 'type' field in property type data")
+        if property_type not in _PROPERTY_TYPE_CLASS_BY_TYPE:
+            raise ValueError(
+                f"invalid property type '{property_type}'. Expected one of {humanize_collection(_PROPERTY_TYPE_CLASS_BY_TYPE.keys(), bind_word='or')}"
+            )
+        cls_ = _PROPERTY_TYPE_CLASS_BY_TYPE[property_type]
+        return cast(Self, cls_.model_validate(data))
+
+    @model_serializer(mode="wrap", when_used="always", return_type=dict)
+    def serialize_property_type(self, handler: SerializerFunctionWrapHandler) -> dict:
+        serialized_data = handler(self)
+        return serialized_data
+
+
+class TextProperty(PropertyTypeDefinition):
+    _property_type = "text"
+    type: Literal["text"] = "text"
+    list: bool | None = Field(
+        default=None,
+        description="Specifies that the data type is a list of values.",
+    )
+    max_list_size: int | None = Field(
+        default=None,
+        description="Specifies the maximum number of values in the list",
+    )
```

Comment:
Suggest creating a `ListableProperty` with `list` and `max_list_size` and use that as a base class for all properties that can become listable.